pkg/agent/clean/active_directory.go

```diff
@@ -0,0 +1,972 @@
+	adGUIDMigrationLabel      = "ad-guid-migration"
+	adGUIDMigrationAnnotation = "ad-guid-migration-data"
+	migratedLabelValue        = "migrated"
+)
+
+type migrateUserWorkUnit struct {
+	distinguishedName    string
+	guid                 string
+	originalUser         *v3.User
+	duplicateUsers       []*v3.User
+	guidCRTBs            []v3.ClusterRoleTemplateBinding
+	duplicateLocalCRTBs  []v3.ClusterRoleTemplateBinding
+	guidPRTBs            []v3.ProjectRoleTemplateBinding
+	duplicateLocalPRTBs  []v3.ProjectRoleTemplateBinding
+	guidTokens           []v3.Token
+	duplicateLocalTokens []v3.Token
+}
+
+type missingUserWorkUnit struct {
+	guid           string
+	originalUser   *v3.User
+	duplicateUsers []*v3.User
+}
+
+type skippedUserWorkUnit struct {
+	guid         string
+	originalUser *v3.User
+}
+
+type LdapErrorNotFound struct{}
+
+// Error provides a string representation of an LdapErrorNotFound
+func (e LdapErrorNotFound) Error() string {
+	return "ldap query returned no results"
+}
+
+type LdapConnectionPermanentlyFailed struct{}
+
+// Error provides a string representation of an LdapConnectionPermanentlyFailed
+func (e LdapConnectionPermanentlyFailed) Error() string {
+	return "ldap search failed to connect after exhausting maximum retry attempts"
+}
+
+type LoginDisabledError struct{}
+
+// Error provides a string representation of an LdapErrorNotFound
+func (e LoginDisabledError) Error() string {
+	return StatusLoginDisabled
+}
+
+func scaledContext(restConfig *restclient.Config) (*config.ScaledContext, error) {
+	sc, err := config.NewScaledContext(*restConfig, nil)
+	if err != nil {
+		logrus.Errorf("[%v] failed to create scaledContext: %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+
+	sc.UserManager, err = common.NewUserManagerNoBindings(sc)
+	if err != nil {
+		logrus.Errorf("[%v] failed to create sc.UserManager: %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+
+	sc.ClientGetter, err = proxy.NewClientGetterFromConfig(*restConfig)
+	if err != nil {
+		logrus.Errorf("[%v] failed to create sc.ClientGetter: %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+	return sc, nil
+}
+
+func adConfiguration(sc *config.ScaledContext) (*v3.ActiveDirectoryConfig, error) {
+	authConfigs := sc.Management.AuthConfigs("")
+	secrets := sc.Core.Secrets("")
+
+	authConfigObj, err := authConfigs.ObjectClient().UnstructuredClient().Get("activedirectory", metav1.GetOptions{})
+	if err != nil {
+		logrus.Errorf("[%v] failed to obtain activedirecotry authConfigObj: %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+
+	u, ok := authConfigObj.(runtime.Unstructured)
+	if !ok {
+		logrus.Errorf("[%v] failed to retrieve ActiveDirectoryConfig, cannot read k8s Unstructured data %v", migrateAdUserOperation, err)
+		return nil, err
+	}
+	storedADConfigMap := u.UnstructuredContent()
+
+	storedADConfig := &v3.ActiveDirectoryConfig{}
+	err = mapstructure.Decode(storedADConfigMap, storedADConfig)
```

Comment:
Reviewing the commit history, it looks like the `common.Decode` refactor was recently enough that it didn't make it into 2.7.5. Because pulling that in would also pull in unrelated changes that are not slated for 2.7.6, we're deciding to leave it as-is.